core/modules/drr.cc

```diff
@@ -299,77 +328,34 @@ void DRR::RemoveFlow(Flow* f) {
   delete f;
 }
 
-llring* DRR::AddQueue(uint32_t slots, int* err) {
-  int bytes = llring_bytes_with_slots(slots);
-  int ret;
-
-  llring* queue = static_cast<llring*>(aligned_alloc(alignof(llring), bytes));
-  if (!queue) {
-    *err = -ENOMEM;
-    return nullptr;
-  }
-
-  ret = llring_init(queue, slots, 1, 1);
-  if (ret) {
-    std::free(queue);
-    *err = -EINVAL;
-    return nullptr;
-  }
-  return queue;
-}
 
 void DRR::Enqueue(Flow* f, bess::Packet* newpkt, int* err) {
   // if the queue is full. drop the packet.
-  if (llring_count(f->queue) >= max_queue_size_) {
+  if (f->queue->Size() >= max_queue_size_) {
     bess::Packet::Free(newpkt);
     return;
   }
 
   // creates a new queue if there is not enough space for the new packet
   // in the old queue
-  if (llring_full(f->queue)) {
+  if (f->queue->Full()) {
     uint32_t slots =
-        RoundToPowerTwo(llring_count(f->queue) * kQueueGrowthFactor);
-    f->queue = ResizeQueue(f->queue, slots, err);
-    if (*err != 0) {
+        RoundToPowerTwo(f->queue->Size() * kQueueGrowthFactor);
+    *err = f->queue->Resize(slots);
+    if (*err) {
```

Comment:
The code from the last PR did not indicate that the queue needed to survive a failed `Resize` operation. In general I would have assumed a queue for which `Resize` failed is no longer valid, in which case you should remove the flow to avoid bad accesses further down the line. Also it might be useful to log something so we know things have gone bad.